### ISSUE TITLE
Iterative consumption of flows

### ISSUE THREAD

**Issue by lowasser** [Kotlin/kotlinx.coroutines#3274]
This is the product of some library design work we've been doing at Google, and we wanted to take this upstream in case there was interest.  The `Flow` API is almost exclusively "push"-based, not "pull" based.  The main alternative approach is to adapt it to a `Channel`, but that often introduces inconvenient concurrency issues (arbitrary buffering, the risk of forgetting to cancel the channel).

What we want is more or less a scoped iteration method.  Here is a sketch of some examples that we think would be better with the proposed API than what `Flow` can do today.

```kotlin
suspend fun <T> Flow<T>.all(predicate: suspend (T) -> Boolean): Boolean = iterate {
  while (hasNext()) {
    if (!predicate(next()) return@iterate false
  }
  true
}

fun <T> Flow<T>.pairElements(): Flow<Pair<T, T>> = flow {
  iterate {
    while (hasNext()) {
      val a = next()
      if (!hasNext()) break
      val b = next()
      emit(Pair(a, b))
    }
  }
}

// this API sort of generalizes e.g. https://github.com/cashapp/turbine
fun testStreamingRpcResponse() = runTest {
  val responses = stub.doRpc(request)
  responses.iterate {
    assertThat(next()).isEqualTo(response1)
    doStimulus(1)
    assertThat(next()).isEqualTo(response2)
    doStimulus(2)
    assertThat(hasNext()).isFalse()
  }
}
```

In short, what we're proposing is

```kotlin
suspend fun <T, R> Flow<T>.iterate(block: ??Iterator<T>.() -> R): R
```

It's not precisely clear what the `??Iterator` should be: we might reuse `ChannelIterator`, but we're not particularly fond of the requirement to call `hasNext()` in the test case especially, and of course this isn't actually a `Channel`.  For ourselves, we are inclined to introduce a separate `FlowIterator` type.

We have an implementation that does not introduce undue concurrency.  To be a little more specific, it does not continue the backing `Flow.collect` until the the next call to `hasNext`, etc.; it's a linear transformation in the same way as `map` and `filter` but not `buffer` or `conflate`.  We can PR it if there's interest.

**Comment by lowasser** [comment]
(To clarify, in the "parsing streams of bytes" use case, we are returning a `Flow` of the parsed messages.)

**Comment by qwwdfsad** [comment]
Hi, I'm still processing your input and trying to internalize all the things in order to further weigh the pros and cons of the solution. Let's discuss individual details in order to be on the same page.

>We don't mind if it's slow
>...
>we would be excited to use this API even if it represented a 30x slowdown, even in production, even if it's only to make code easier to follow

It's good to hear that, meaning that we have a bit more space in our design. But it's important to remember that we have plenty of other consumers, including, but not limited to, Android developers, server-side developers, high-performance web-frameworks, high-throughput backend apps and IDEs. Having two similar APIs that, at the first glance, differs only in code-style preferences, but one being 30 times slower, will bring **irreversible** damage to the ecosystem in the long term, especially when both pull and push based APIs are equally appealing. 

##

>
```
suspend fun <T> Flow<T>.all(predicate: suspend (T) -> Boolean): Boolean = iterate {
  while (hasNext()) {
    if (!predicate(next()) return@iterate false
  }
  true
}
```

This is an example of what can bring potential harm. One can implement it via `collectWhile` using a pull-based model, and one can implement it using `iterate`, but being an **order of magnitude** slower. While being acceptable in some cases, it eventually will separate the code style, will bring the debates over code-style and a lot of articles "Do and dont's for performance" (and the veil of mystery around the mythical abstract "performance" won't make it any easier for us).

## 

>https://github.com/Kotlin/kotlinx.coroutines/pull/3278

Would it be okay for you to semi-finish the PR? It is a decent first peek at the API shape, but unfortunately, it doesn't compile and doesn't have any tests. It would be really nice to see API and tests that you _already have written_ to give us a taste of how you are leveraging it. There are some examples, but they have their own pitfalls. Real-world use-cases are always better :)

##

>This is a very simple state machine, but it's much clearer to write
```
fun Flow<T>.pairElements() = flow {
  iterate {
    while (hasNext()) {
      emit(Pair(next(), next()))
    }
  }
}
```

This particular example looks really concise, works and passes basic testing. As long as the original flow emits an even number of elements, otherwise, it throws `NoSuchElementException`.
Of course, channels and collection's iterators have the same pitfall, but their use pattern is completely different -- their iterators are almost never expected to be used directly, but rather with `for`-loop convention, `consume` extension or stdlib primitives. Here we explicitly delegate this responsibility to end-users, potentially creating the space for hard-to-spot bugs that type system cannot prevent.

##

>A fully sequential, single-threaded dispatcher prevents "races" in the sense of JVM races, but doesn't solve the "race" in the sense that the dispatcher chooses which of two eligible coroutines get resumed first, and debugging requires the dispatcher to make a specific choice there that a dispatcher can't guess.


Thanks for the control flow diagram, it really made the picture clear. I've immediately started thinking about more intrusive `UNDISPATCHED` primitive for this purpose and noticed your comment

>We've been hesitant to fully adopt Turbine because it dallies with UNDISPATCHED, which we've seen lead to impossible-to-debug issues

Could you please elaborate on that and show a few examples of such issues?

##

I'm looking forward to seeing more examples (I suppose the parsing is the most interesting and hard-to-emulate one) to see if there are ways to improve the API and get around potential drawbacks

**Comment by JakeWharton** [comment]
Do you mean `Unconfined`? Turbine's use of `UNDISPATCHED` merely starts the upstream flow collection synchronously.

**Comment by lowasser** [comment]
I don't have specific issues I can refer to about Unconfined or UNDISPATCHED, though both make us extremely nervous.  Certainly in Java land we've had lots of issues caused by directExecutor.  We specifically haven't seen any such issues with Turbine, but then, we have it on a tight allowlist just because we don't trust that there aren't issues with Unconfined and UNDISPATCHED too subtle for us to notice.  Similarly, we would be unhappy about more intrusive primitives and their edge cases.

We care about performance, to be clear, but we think race conditions in common patterns that are subtle enough that they're not obvious to you or Roman at first glance are pretty damaging to the ecosystem themselves.

I've committed a working version with some of our tests to the PR, though it doesn't have as many tests as I'd like yet.  I've also modified it to avoid the use of channels, instead just passing continuations back and forth.  I'm less confident in its correctness, though.

(Possibly an advantage of the API in the current PR: it can't actually be for-eached over, which makes it more likely to be used in iterator-specific ways?)

**Comment by JakeWharton** [comment]
Turbines use of unconfined is an optimization to avoid double dispatch similar to one of your goals here. It isn't strictly required, and runTest seems to have broken its behavior so probably worth removing.

**Comment by jingibus** [comment]
It's not just Turbine-type interactions that have need fo "pull" consumption. We also need it for interaction with fakes.

Internally at Cash, we've been attacking this problem by using a "pull" consumption tool that Kotlin already has: `Channel`. 

We facade `Channel`'s API with one better suited to our test requirements. Here's a simplified sketch of that API:

```kotlin
class AwaitChannel<T> {
  private val channel = Channel<T>(capacity = UNLIMITED)

  fun add(value: T) { channel.trySend(value) }

  val isEmpty: Boolean get() { return channel.isEmpty }

  /**
   * Wait for a value on this channel for 1s, throwing a [CancellationException] if none arrives.
   */
  suspend fun <T> awaitValue() = try {
    withTimeout(1000L) {
      channel.receive()
    }
  }

  /**
   * Return a value, or null if empty. 
   *
   * To be used from legacy, non-suspendful call sites. Will throw
   * when called from a non-suspending context.
   */
  fun <T> takeValue(): T? {
    assertCallingContextIsNotSuspended()
    return channel.tryReceive().getOrNull()
  }
```

This is essentially the same thing Turbine provides, but in a form that works in our Fakes.

I think the "it's just a channel" idea can work in Turbine, too. I did some homework on this last week, and haven't yet found a major roadblock.

**Comment by lowasser** [comment]
Does that approach address the race issue discussed in the PR [here](https://github.com/Kotlin/kotlinx.coroutines/pull/3278#issuecomment-1125513103)?

**Comment by jingibus** [comment]
We have only been concerned with races in test. Test has firm expectations of consistency: many immaterial changes in order may occur in multithreaded prod without harm, but immaterial changes in ordering under test will result in flaky test failures.

Since races are introduced by threading, not coroutines, we police dispatchers aggressively:

* All dispatchers are injected as `EmptyCoroutineContext` in test
* All Rx schedulers are injected as `Schedulers.trampoline()` in test
* We stay on guard against problematic Kotlin APIs that silently introduce `Dispatchers.Default` (e.g. `rx(Single|Maybe|Completable|Observable)`, dispatcherless `CoroutineScope()`)

**Comment by lowasser** [comment]
There are plenty of races that can happen as a result of coroutines and not threading!  https://github.com/Kotlin/kotlinx.coroutines/pull/3278#issuecomment-1124141560 is an example of code that nondeterministically throws or succeeds, even run on a single-threaded dispatcher.

**Comment by jingibus** [comment]
> https://github.com/Kotlin/kotlinx.coroutines/pull/3278#issuecomment-1124141560 is an example of code that nondeterministically throws or succeeds, even run on a single-threaded dispatcher.

I've stewed on this claim for a few weeks, and reviewing the discussion I think that it doesn't really hit at anything I really care about. And that is because I truly care about *actual* nondeterminism: identical code that yields different results on different invocations. 

The following paragraph I think makes the different issues pretty clear:

> A clever dispatcher doesn't solve this problem, though! https://github.com/Kotlin/kotlinx.coroutines/pull/3278#issuecomment-1125513103 shows the control flow diagrams that are the root of the issue, which is that two separate coroutines are both eligible to run at the same point in the program, as far as any dispatcher is concerned, but we don't want one of them to run (yet) when debugging -- and the dispatcher API means that it can't possibly know which one we want. A fully sequential, single-threaded dispatcher prevents "races" in the sense of JVM races, but doesn't solve the "race" in the sense that the dispatcher chooses which of two eligible coroutines get resumed first, and debugging requires the dispatcher to make a specific choice there that a dispatcher can't guess.

So we have two issues: JVM races, and dispatcher races.

I agree that these are both problems: as a programmer writing a test, I care quite a bit about dispatcher races. Dispatcher races make it difficult to understand what the outcome of my test code will be, and that makes test writing very finicky.

But: that is a different order of problem from a *JVM* race. As someone who tends to get pinged about test issues in my organization, I don't generally get out of bed for a dispatcher race, because it is a problem localized to that engineer. 

If the engineer is experiencing a JVM race, on the other hand, that is a major issue. JVM races can expose every single engineer in the team to flaky CI tests. That's when I zoom in and apply the three rules given above to root cause the problem.

So: our language is letting us down here. If I say "race", I may be referring to a dispatcher race (which is a conceptual race), or a JVM race (which is an actual race). I admit to using "race" myself interchangeably for these two very ideas. It's definitely a cause of discord: anytime you say "race condition," engineers run screaming.

And they should! But maybe sometimes faster than others.

**Comment by lowasser** [comment]
I agree that these are two different sorts of problem, but I'm not actually convinced of your relative ordering of them.  Here are some counterpoints I'd like to make:

* JVM races can be caught by instrumentation such as [TSAN](https://wiki.openjdk.org/display/tsan/Main), but dispatcher races cannot.
* Dispatcher races can go the other way whenever you attempt to upgrade coroutines -- which is constantly necessary, given the rate of bugfixes.  This can result in large numbers of subtle test breakages that coroutines maintainers in our organization need to resolve before pushing upgrades.
* Dispatcher races in production code can manifest in the tests of any team using that code.
* Test-driven development is especially badly effected when one can't actually predict the result of such a race.

But finally, I think it comes down to the issue that a developer should be able to predict the deterministic behavior of this code in advance.  There's no good reason for it to be hard to predict.  Worse, it _looks_ predictable and isn't.

**Comment by jingibus** [comment]
Those arguments make sense to me. But even if I grant all of them, is the conclusion (dispatcher races are dangerous) actionable? Even with the proposed `iterate`?

E.g. consider an operation like `combine`:

```kotlin
fun <T> combine(sourceFlows: List<Flow<T>) = flow {
  val items = Channel<T>()
  coroutineScope {
    sourceFlows.forEach { 
      launch { it.collect { item -> items.send(item) } }
    }
    items.consumeEach { emit(it) }
  }
}
```

For this flow, the order of items emitted depends on the order in which the `sourceFlows` collections are dispatched. So this is a dispatcher race. And a bread and butter one for us: our async business logic frequently collates data from multiple sources.

How do we get around that? A tool like `iterate` doesn't actually address this race, right? `iterate` punctures the `suspend fun collect` black box, but there are boxes within boxes. 

Here's my POV: there is no silver bullet to make any coroutines code conceptually "deterministic," because there's none for concurrent code in general. Programming coroutines code necessarily means programming concurrently, and thus programming in a world where dispatchers will race. This is why the whole coroutines/channels/jobs toolkit exists: to provide a toolset for connecting concurrent processes with one another that can be reasoned about, even in the presence of actual nondeterminism.

But actual nondeterminism is a huge problem under test. (I know; I've tried it, intrepid fool that I am.) We have to rely on the JVM determinism of a single threaded dispatcher to patch things over that last bit, and eat the issues with coroutines upgrades. (We're a lot smaller than you; I shudder to think of the pain you're experiencing there)

So my approach has been to say, "Coroutines tests are just coroutines code. That means they're concurrent. That means learning new techniques to govern the concurrent flow of data and execution in your tests (e.g. use channels to build fakes that you can interact with, and don't use shared state to communicate with test code). It also means abiding by a few rules about dispatchers to avoid flakiness on CI. Do this, and you'll find yourself more fluent in your test code and in prod, since you'll be using the same basic toolset in both places."

I've personally been pretty happy with the result from a TDD perspective. I'd love to see an even better alternative, of course, but until that happens this has been quite effective for us.
